Mathlib/Lean/Meta/RefinedDiscrTree.lean

```diff
@@ -0,0 +1,1204 @@
+    unless fnType matches .forallE .. do
+      fnType ← whnfD (fnType.instantiateRevRange j i args)
+      j := i
+    let .forallE _ d b bi := fnType | throwError m! "expected function type {indentExpr fnType}"
+    fnType := b
+    result := result.push (← isIgnoredArg args[i]! d bi)
+  return result
+where
+  /-- Determine whether the argument should be ignored. -/
+  isIgnoredArg (arg domain : Expr) (binderInfo : BinderInfo) : MetaM Bool := do
+    if domain.isOutParam then
+      return true
+    match binderInfo with
+    | .instImplicit => return true
+    | .implicit
+    | .strictImplicit => return !(← isType arg)
+    | .default => isProof arg
+
+
+
+/-- Introduce new lambdas by η-expansion. -/
+@[specialize]
+def etaExpand (args : Array Expr) (type : Expr) (lambdas : List FVarId) (goalArity : Nat)
+    (k : Array Expr → List FVarId → MetaM α) : MetaM α  := do
+  if args.size < goalArity then
+    withLocalDeclD `_η type fun fvar =>
+      etaExpand (args.push fvar) type (fvar.fvarId! :: lambdas) goalArity k
+  else
+    k args lambdas
+termination_by goalArity - args.size
+
+
+/-- Normalize an application of a heterogenous binary operator like `HAdd.hAdd`, using:
+- (1) `f = fun x => f x` to increase the arity to 6
+- (2) `(f + g) a = f a + g a` to try to decrease the arity to 6
+- (3) `(fun x => f x + g x) = f + g` to get rid of the lambdas in front -/
+def reduceHBinOpAux (args : Array Expr) (lambdas : List FVarId) (instH instPi : Name) :
+    MetaM (Option (Expr × Expr × Expr × List FVarId)) := do
+  let some (mkApp2 (.const instH' _) type inst) := args[3]? | return none
+  unless (instH == instH') do return none
+  some <$> do
+  if args.size ≤ 6 then
+    -- (1)
+    etaExpand args type lambdas 6 fun args lambdas =>
+      distributeLambdas lambdas type args[4]! args[5]!
+  else
+  -- (2)
+  let mut type := type
+  let mut inst := inst
+  let mut lhs := args[4]!
+  let mut rhs := args[5]!
+  for arg in args[6:] do
+    let mkApp3 (.const i _) _ f inst' := inst | return (type, lhs, rhs, lambdas)
+    unless i == instPi do return (type, lhs, rhs, lambdas)
+    type := .app f arg
+    inst := inst'
+    lhs := .app lhs arg
+    rhs := .app rhs arg
+  distributeLambdas lambdas type lhs rhs
+where
+  /-- (3) `(fun x => f x + g x) = f + g` to get rid of the lambdas in front -/
+  distributeLambdas (lambdas : List FVarId) (type lhs rhs : Expr) :
+      MetaM (Expr × Expr × Expr × List FVarId) := match lambdas with
+    | fvarId :: lambdas => do
+      let decl ← fvarId.getDecl
+      let type := .forallE decl.userName decl.type (type.abstract #[.fvar fvarId]) decl.binderInfo
+      let lhs  := .lam     decl.userName decl.type (lhs.abstract  #[.fvar fvarId]) decl.binderInfo
+      let rhs  := .lam     decl.userName decl.type (rhs.abstract  #[.fvar fvarId]) decl.binderInfo
+      distributeLambdas lambdas type lhs rhs
+    | [] => return (type, lhs, rhs, [])
+
+/-- Normalize an application if the head is  `+`, `*`, `-` or `/`.
```

Comment:
I think so. In those we have the rules `fun i => s • x i` = `s • x` and `fun i => x i ^ b` = `x ^ b`. I think I'll move this part of the code to a separate file (the part for normalizing pi instances). In the long run when `RefinedDiscrTree` will move to std/Boost, this bit of code needs to stay in Mathlib because the pi instances are defined here. Then there should be a way to dynamically extend the reduction function. Then this code could go into `Mathlib.Algebra.Group.Pi.basic`, do it would run this if and only if the pi instances are imported.

Reply:
I have reimplemented this code now, adding support for HSMul, HVAdd and HPow. In the process I also discovered a bug in an edge case. I have now added a test file to see that it is indeed doing what it should. There is now the function `reducePi` which does this normalization returning the result as an `Expr`, so that it can also be used somewhere else if that is useful.